### ISSUE TITLE
Coolermaster support

### ISSUE THREAD

**Issue by phomes** [libratbag/libratbag#487]
I am working on a driver for Coolermaster MM520 and MasterMouse S. It seems @elFarto is too. Let's see what we can share :)

**Comment by elFarto** [comment]
I'm currently implementing a Coolermaster CMStorm Recon. I've pushed the latest version to a branch on my fork.

**Comment by phomes** [comment]
It looks like you have made quite a lot of progress. You are much further along than I am. I do not think that the devices speak the same protocol though. On my device the settings are sent as a storm of reports, each setting a specific value (e.g. one report for led mode, one for led colors, one for report rate, etc). It seems that your devices transfers full profiles instead.
So I encourage you to continue with your separate driver, as it seems that little can be shared.

**Comment by elFarto** [comment]
It might be worth me renaming the driver to cmstorm, rather than coolermaster. That said, I think it's possible to change individual settings if craft the 0xD1 command correctly (it just seems to be a generic write to memory command). If you could upload a trace of the commands I can confirm if they're different. I'm sending full profiles because that's what the Windows software does.

However, given the "quality" of the firmware/software for this mouse, I wouldn't be surprised if they were creating everything from scratch for each different mouse. Looks like the Recon, Spawn and Xornet all share an MCU so they can probably share the same driver.

**Comment by verstatx** [comment]
Just a heads up, I think the CM Recon driver should be renamed to be related to _LaVIEW Technology_ (the Hong Kong company).

I have a Recon and a Mionix Avior 7000, both reference Laview Technology in xinput, and both use the same system of full profile uploads in order to change settings. Additionially, the windows software for both is practically the same. I assume that Laview was responsible for the software/firmware and considering they have developed mice for different brands (in addition to their own mice), I would assume there are other mice from other brands using the same approach.

I haven't looked at elFarto's code yet, and I don't have access to the Recon at the moment, but when I do, I would be happy to help out with the driver.

**Comment by verstatx** [comment]
I was finally able to look into this and the protocols seem way different. Notably the size of a profile page uploaded to the mouse is 1041 bytes vs 65, the command to write a profile on the Avior starts with `0x61 08` as opposed to `0x03 d1`, and each setting appears to be delimited by `0x0f`. About the only similarity I could find at a first glance was the button type, but even the notation following that is incompatible. Do note that this was a cursory glance and that I did not look into macros at all. Here's a small pcap of applying the Avior's settings: [mionix-avior-apply-settings.pcapng.gz](https://github.com/libratbag/libratbag/files/2386193/mionix-avior-apply-settings.pcapng.gz). The software uploads 5 profiles at a time, each with at least(?) 2 pages, empty or not.

So while the software and and profile file formats were very similar, the devices themselves do not look very compatible. Whatever few similarities there are in the protocol, working around the incompatibilities is probably not worth it.

**Comment by elFarto** [comment]
I agree, It certainly looks like it's following the same basic design pattern, but I don't think it's similar enough to warrant having one driver covering both protocols.

**Comment by whot** [comment]
without knowing anything about the protocol, it sounds like it's going to be more effective writing a separate driver first, then seeing whether it's feasible to merge them after. easier to write this way and easier to test too.

**Comment by telephrag** [comment]
Has this gone anywhere? It seems like everyone involved deleted all related code from their profiles.
